lib/openssh/sshd.go

```diff
@@ -0,0 +1,210 @@
+/*
+Copyright 2023 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package openssh
+
+import (
+	"fmt"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/teleport/lib/auth"
+	"github.com/gravitational/trace"
+)
+
+const sshdBinary = "sshd"
+
+const sshdConfigSectionModificationHeader = "# Created by 'teleport join openssh', do not edit"
+
+var (
+	// AgentlessSSHDConfigPath is the path to write teleport specific SSHD config options
+	AgentlessSSHDConfigPath = "/etc/teleport/sshd.conf"
+	AgentlessKeysDir        = "/etc/teleport/openssh"
+	// agentlessSSHDConfig is Include directive added to the sshd_config file
+	agentlessSSHDConfigInclude = "Include " + AgentlessSSHDConfigPath
+)
+
+const (
+	TeleportKey       = "teleport"
+	TeleportCert      = "teleport-cert.pub"
+	TeleportOpenSSHCA = "teleport_user_ca.pub"
+)
+
+type sshdBackendOperations interface {
+	restart() error
+	checkConfig(path string) error
+}
+
+type SSHd struct {
+	sshd sshdBackendOperations
+}
+
+func NewSSHD() SSHd {
+	return SSHd{
+		sshd: &sshdBackend{},
+	}
+}
+
+func (b *SSHd) UpdateConfig(u SSHdConfigUpdate, restart bool) error {
+	needsUpdate, err := checkSSHDConfigAlreadyUpdated(u.SSHdConfigPath, agentlessSSHDConfigInclude)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	if needsUpdate {
+		if err := prefixSSHDConfig(u.SSHdConfigPath, agentlessSSHDConfigInclude); err != nil {
+			return trace.Wrap(err)
+		}
+	}
+
+	configUpdate := fmtSSHDConfigUpdate(
+		u.OpenSSHCAPath,
+		u.HostKeyPaths,
+		u.HostCertPaths,
+	)
+
+	sshdConfigTmp, err := os.CreateTemp("", "")
+	if err != nil {
+		return trace.ConvertSystemError(err)
+	}
+	defer sshdConfigTmp.Close()
+	if _, err := sshdConfigTmp.Write([]byte(configUpdate)); err != nil {
+		return trace.ConvertSystemError(err)
+	}
+
+	if err := sshdConfigTmp.Sync(); err != nil {
+		return trace.ConvertSystemError(err)
+	}
+
+	if err := b.sshd.checkConfig(sshdConfigTmp.Name()); err != nil {
+		return trace.Wrap(err)
+	}
+
+	if err := os.Rename(sshdConfigTmp.Name(), u.TeleportSSHDConfigPath); err != nil {
```

Comment:
This only works if the `sshdConfigTmp` happened to be created on the same filesystem as `u.TeleportSSHDConfigPath`, which is not guaranteed (often times `/tmp/` will be a tmpfs, for instance). You should create the temporary file in the same directory as `TeleportSSHDConfigPath` and defer its deletion blindly (so that if we exit out before the file is renamed, the temporary file is deleted).